docs/model_classes/usermodel.rst

```diff
@@ -21,17 +21,194 @@ external code.
    (0,0) vs (1,1) sort of thing?"
    
 
+Overview
+========
+
+A single model is defined by the parameters of the model - stored
+as `sherpa.models.parameter.Parameter` instances - and the function that
+takes the parameter values along with an array of grid values to calculate
+values of the model. This is packaged together in a class for each model,
+and instances of this class can then be used to fit data.
+
+Sherpa provides several base classes in `sherpa.models.model`; the most
+commonly used ones are:
+
+* `~sherpa.models.model.ArithmeticModel` is the main base class for deriving user models since
+  it supports combining models (e.g. by addition or multiplication) and
+  a cache to reduce evaluation time at the expense of memory use.
+  (`~sherpa.models.model.ArithmeticConstantModel` and `ArithmeticFunctionModel` are less general
+  than `~sherpa.models.model.ArithmeticModel` and can be used to represent
+ a constant value or a function.)
+
+* `~sherpa.models.model.RegriddableModel1D` and
+  `~sherpa.models.model.RegriddableModel2D` which build on
+  `~sherpa.models.model.ArithmeticModel` to allow a model to be
+  evaluated on a different grid to that requested.
+
+To define a new model, a new class inherits from one of these base classes and
+implements an ``__init__`` method that defines the model parameters
+and a ``calc`` method that, given the parameter values and the coordinates
+at which the model is to be evaluated, returns the model values.
+
+Calculating the model values
+============================
+
+Many Sherpa models are set up to work with both integrated and non-integrated data,
+some might even work with data of different dimensionality. Because of that, the input
+to the ``calc`` method can take several forms:
```

Comment:
So, maybe we should say that the default signature is (or should be, I haven't bothered to actually check `Model.calc`)

```python
def calc(self, p, *args, **kwargs):
```

(or use `pars` but need to check with e.g. #1797) and then it's up to the model to decide what to do - i.e. if you have a 1D model you could say

```python
def calc(self, p, x, xhi=None, **kwargs):
```

One concern I have is that we could also have this as one of

```
def calc(self, p, x0, x1=None, **kwargs):
def calc(self, p, lo, hi=None, **kwargs):
... etc ...
```

and so we (as in Sherpa) need to be careful to make sure we don't use named arguments when evaluating models.

If you want to sound Computer-Sciencey then you can throw around [Liskov Substitution Principle](https://en.wikipedia.org/wiki/Liskov_substitution_principle) but I think our code base is well beyond that.

Reply:
> we (as in Sherpa) need to be careful to make sure we don't use named arguments when evaluating models.

One more reason to change to an interface that gets the data in some more structured form (#1796), but here I'm just writing down what we do now. I'll add a suggestion with the "default" way to write that, but we don't have to enforce it.